framework/filters/AccessRule.php

```diff
@@ -136,19 +142,26 @@ protected function matchController($controller)
      */
     protected function matchRole($user)
     {
-        if (empty($this->roles)) {
+        $privileges = empty($this->roles) ? [] : $this->roles;
+
```

Comment:
these arrays are quite vague. Using objects will improve one's understanding of $roles & $permissions and their mutual relation.